src/stan/model/indexing/lvalue_varmat.hpp

```diff
@@ -306,37 +309,53 @@ inline void assign(
                                assign_rows, name, y.rows());
   stan::math::check_size_match("matrix[multi,multi] assign", "left hand side",
                                assign_cols, name, y.cols());
-  arena_t<std::vector<std::array<int, 2>>> x_idx;
-  arena_t<std::vector<std::array<int, 2>>> y_idx;
-  x_idx.reserve(assign_rows * assign_cols);
-  y_idx.reserve(assign_rows * assign_cols);
-  // Need to remove duplicates for cases like {2, 3, 2, 2}
+  using arena_vec = std::vector<int, stan::math::arena_allocator<int>>;
+  using pair_type = std::pair<int, arena_vec>;
+  arena_vec x_col_idx(assign_cols);
+  arena_vec x_row_idx(assign_rows);
+  std::unordered_set<int> x_set;
+  x_set.reserve(assign_cols);
+  std::unordered_set<int> x_row_set;
+  x_row_set.reserve(assign_rows);
+  arena_t<Eigen::Matrix<double, -1, -1>> prev_vals(assign_rows, assign_cols);
+  // Need to remove duplicates for cases like {{2, 3, 2, 2}, {1, 2, 2}}
+  for (int i = assign_rows - 1; i >= 0; --i) {
+    if (likely(x_row_set.insert(idxs.head_.ns_[i]).second)) {
+      stan::math::check_range("matrix[multi, multi] assign row", name, x.rows(),
+                              idxs.head_.ns_[i]);
+      x_row_idx[i] = idxs.head_.ns_[i] - 1;
+    } else {
+      x_row_idx[i] = -1;
+    }
+  }
   for (int j = assign_cols - 1; j >= 0; --j) {
-    for (int i = assign_rows - 1; i >= 0; --i) {
-      if (!internal::check_duplicate(x_idx, idxs.head_.ns_[i] - 1,
-                                     idxs.tail_.head_.ns_[j] - 1)) {
-        y_idx.push_back(std::array<int, 2>{i, j});
-        stan::math::check_range("matrix[multi, multi] assign row", name,
-                                x.rows(), idxs.head_.ns_[i]);
-        stan::math::check_range("matrix[multi, multi] assign col", name,
-                                x.cols(), idxs.tail_.head_.ns_[j]);
-        x_idx.push_back(std::array<int, 2>{idxs.head_.ns_[i] - 1,
-                                           idxs.tail_.head_.ns_[j] - 1});
+    if (likely(x_set.insert(idxs.tail_.head_.ns_[j]).second)) {
```

Comment:
For multi multi overload you need to deduplicate both indices at the same time - by creating a set of pairs instead of two sets. This code would also "deduplicate" cases like `{2,2,3}, {2,3,3}`.

Reply:
Yeah so this one took me a while to figure out but this is actually safe. You can checkout the `multi_multi_matrix` test here to see that I test against "deduping" against a set of pairs.

The important thing to note is that for something like `{1, 2, 3}, {4, 5, 6}` you can read it as "for column 4 assign to cells 1, 2, 3. For column 5 assign to cells 1, 2, 3, for column six...". So the rows assignment indices are reused for each column. That let's us just deduplicate the row assignment vector once

Reply:
Right, Stan has a bit different understanding of 2D indexing than I do. This is fine.